### PR TITLE
reimported should detect aliased imports

### LINKED ISSUE #4836: False negative ``reimported`` when using an import alias
### Current problem

I think it would be nice if Pylint could check if there is same import name from different packages.
For example, in the code below, import order is fine, but the 'Path' imports overlap, causing errors somewhere.
```python
from pathlib import Path
from FastAPI import Path
```

### Desired solution

overlapping imports found. `Path`

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #4836

Issue brought up the fact that the `reimported` message was confusing and hinted at something else, while also giving us more test cases to improve the check for imports using aliases.

### CODE REVIEW

File: tests/functional/r/reimported.py
Comment (Pierre-Sassoulas) [inline tests/functional/r/reimported.py:2]:
```suggestion
# pylint: disable=missing-docstring,unused-import,import-error, wildcard-import,unused-wildcard-import,redefined-builtin,no-name-in-module
# pylint: disable=ungrouped-imports, wrong-import-order,wrong-import-position,consider-using-from-import
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
LGTM, I would go further and rename ``reimported`` to something like ``import-name-clash``or something clearer about what the problem is. "reimported" feel like what should be raised for:
```pyhon
from pathlib import Path
from pathlib import Path
```
Which is a lot less problematic than the problem we detect here.

**Comment by Pierre-Sassoulas** [comment]
I realize by reading the issue that we already had this discussion. Now that I think about it again this feel like a new check, which should have a different message and probably also a different level (warning at least, while reimported is probably just a convention).

**Comment by clavedeluna** [comment]
I followed the conclusion of the issue. If maintainers would like to change scope, lmk I will close this.

**Comment by Pierre-Sassoulas** [comment]
Let's pursue the discussion in the original issue. I feel that your implementation would permit to separate the two use case relatively easily though.

**Comment by Pierre-Sassoulas** [comment]
Let's separate the two checks between ``reimported`` and ``shadowed-import``. I also think that this is a false positive for ``shadowed-import`` as the import is not shadowed:
```python
from pandas._libs import (
    algos as libalgos,
    hashtable as ht,
)
import pandas.core.algorithms as algos
```

But this should raise shadowed-import:

```python
from pandas._libs import (
    algos,
    hashtable as ht,
)
import pandas.core.algorithms as algos  # [shadowed-import]
```

And this should raise reimported:

```python
from pandas._libs import (
    algos as libalgos,
    hashtable as ht,
)
import pandas._libs.algos as algos  # [reimported]
```

**Comment by clavedeluna** [comment]
As a principle, I don't like changing the scope of a PR after work that _could_ be merged is complete. This PR could be merged as is, and through an iterative process, a new issue can be opened to then separate the two checks. Without this, you're asking a contributor to do more work than they initially signed up for in an issue. So not saying scope can't change -it always will. Just saying that we should manage it by cutting it with issues.

**Comment by Pierre-Sassoulas** [comment]
Sorry about that but I don't think this is mergeable as is. This would make ``reimported`` confusing by adding a new use case that does not correspond to the current name and we'd have to make real breaking change to correct it later if we release it. (a user disable reimported somewhere and the message then change to import-shadowed and she has to disable it again)  So if we merge this it's going to make the main branch un-realeasable until someone fix it and that's not something I want to have to keep in mind.

File: pylint/checkers/imports.py
Comment (clavedeluna) [inline pylint/checkers/imports.py:938]: I see a few similar is_message_enabled` checks in the codebase, should these actually be replaced with `@only_required_for_messages(...`?

Comment (Pierre-Sassoulas) [reply pylint/checkers/imports.py:938]: I think ``only_required_for_messages`` only work for visitor function, so ``is_message_enabled`` is a suboptimization for non visitor function. Imo it's premature optimization to add them when implementing but also making the code worst to remove them so we should just leave it alone unless there's an actual problem with it.

File: doc/whatsnew/fragments/4836.bugfix
Comment (Pierre-Sassoulas) [inline doc/whatsnew/fragments/4836.bugfix:1]:
```suggestion
Update ``reimported`` help message for clarity and add a ``shadowed-import`` message for aliased imports.
```

File: pylint/checkers/imports.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/imports.py:322]:
```suggestion
        "Used when a module is aliased with a name that shadows another import.",
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks pretty great 👍 Some minor nitpicks and let's merge !

**Review by Pierre-Sassoulas — APPROVED** [review]

File: doc/whatsnew/fragments/4836.bugfix
Comment (Pierre-Sassoulas) [reply doc/whatsnew/fragments/4836.bugfix:1]: https://github.com/PyCQA/pylint/pull/7903

**Pull request merged**
